### PR TITLE
SDK - Added pinned dependency snapshot

### PR DESCRIPTION
**Pull request by Ark-kun**
Issue: https://github.com/kubeflow/pipelines/issues/3078

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/3303)

### CODE REVIEW

File: sdk/python/setup.py [hunk 1]
```diff
@@ -20,25 +20,21 @@
...
-    'six >= 1.10',
-    'certifi',
     'python-dateutil',
     'PyYAML',
     'google-cloud-storage>=1.13.0',
-    'kubernetes>=8.0.0, <=10.0.0',
-    'PyJWT>=1.6.4',
-    'cryptography>=2.4.2',
+    'kubernetes>=8.0.0, <11.0.0',
+    #'PyJWT>=1.6.4',
```
Comment (Bobgy) [inline sdk/python/setup.py]: nit: why is this commented out?

File: sdk/python/requirements.txt [hunk 2]
```diff
@@ -1,19 +1,48 @@
...
-jsonschema >= 3.0.1
-tabulate == 0.8.3
-click == 7.0
-Deprecated
-strip-hints
+#
+# This file is autogenerated by pip-compile
+# To update, run:
+#
+#    pip-compile --output-file=requirements.txt requirements.in
```
Comment (Bobgy) [inline sdk/python/requirements.txt]: Previously, we could only stably update requirements.txt using docker: https://github.com/kubeflow/pipelines/pull/3161/files#diff-4a38caa8c75b104ebd2a31a36f5c3b44R4

Looks like different python version will have different set of available dependencies, so generated requirements.txt can be different.
I'm not very familiar with which python versions our sdk needs to support, will this work for sdk?

**Comment by numerology** [comment]
If I understand it correctly, `requirements.txt` is for testing purpose only, right?

File: sdk/python/setup.py (hunk 1 above)
Comment (Ark-kun) [reply sdk/python/setup.py]: I'm not sure this package is used.

File: sdk/python/requirements.txt (hunk 2 above)
Comment (Ark-kun) [reply sdk/python/requirements.txt]: My impression was that for KFP the packages are the same regardless of the python version. Do you have a specific example of difference?
In Travis we test 2.5, 3.6, 3.7 and want to test 3.8.

**Comment by Ark-kun** [comment]
> If I understand it correctly, `requirements.txt` is for testing purpose only, right?

It can also be useful for reproducible "known good" install.

File: sdk/python/requirements.txt (hunk 2 above)
Comment (Bobgy) [reply sdk/python/requirements.txt]: https://github.com/kubeflow/pipelines/pull/3161#issuecomment-590986792
The image was using python 3.5, probably it hasn't been supported by some new libraries.

Comment (Ark-kun) [reply sdk/python/requirements.txt]: I found one package (zipp) that has dropped support for python3.5.
So it makes sense to create the snapshot using the lowest version of python that we support. The snapshot will be tested on all supported python version by the presubmit tests.

Comment (Bobgy) [reply sdk/python/requirements.txt]: SGTM

File: backend/Dockerfile [hunk 3]
```diff
@@ -25,11 +25,6 @@ RUN if [ "$use_remote_build" = "true" ]; then \
 
 # Compile
 FROM python:3.5 as compiler
-RUN apt-get update -y && \
-    apt-get install --no-install-recommends -y -q default-jdk python3-setuptools python3-dev
-RUN wget https://bootstrap.pypa.io/get-pip.py && python3 get-pip.py
```
Comment (Bobgy) [inline backend/Dockerfile]: If we skip the step to upgrade pip, we may end up getting previous flakiness caused by old version of pip cannot retry problem.

Comment (Ark-kun) [reply backend/Dockerfile]: Thanks for the information. This is important. 
I will restore upgrading pip.
BTW, do you think we should continue using python:3.5 for the sample compilation? Maybe we should switch to 3.7?

Comment (Bobgy) [reply backend/Dockerfile]: Yep, 3.7 sounds good to me, I remember I tried to install python 3.5 using miniconda and apparently it is no longer supported.

**Comment by Ark-kun** [comment]
@Bobgy The tests are passing now.

File: sdk/python/requirements.txt [hunk 4]
```diff
@@ -1,19 +1,48 @@
...
+google-api-core==1.16.0   # via google-cloud-core
+google-auth==1.11.3       # via -r requirements.in, google-api-core, google-cloud-storage, kubernetes
+google-cloud-core==1.3.0  # via google-cloud-storage
+google-cloud-storage==1.26.0  # via -r requirements.in
+google-resumable-media==0.5.0  # via google-cloud-storage
+googleapis-common-protos==1.51.0  # via google-api-core
+idna==2.9                 # via requests
+importlib-metadata==1.5.0  # via jsonschema
+jsonschema==3.2.0         # via -r requirements.in
+kfp-server-api==0.3.0     # via -r requirements.in
```
Comment (numerology) [inline sdk/python/requirements.txt]: Just curious, why we have `kfp-server-api==0.3.0` while in requirements.in we have `kfp-server-api==0.2.5`?

Comment (Bobgy) [reply sdk/python/requirements.txt]: sounds like we should drop kfp-server-api from requirements.in

Comment (Ark-kun) [reply sdk/python/requirements.txt]: Sorry. It was a bug introduced during resolving merge conflict (someone has reformatted setup.py).
I'll set it to 0.3.0 in both places.

**Pull request merged**